### PR TITLE
impl create update graph gql apis as plugins

### PR DESCRIPTION
**Pull request by shivam-880**
### What changes were proposed in this pull request?
Impl create/update graph gql apis as plugins

### Why are the changes needed?
These apis are custom apis which are local to a use case

### Does this PR introduce any user-facing change? If yes is this documented?
Yes

### How was this patch tested?
Integration tests

### Issues

_If this resolves any issues, please link to them here, the format is a KEYWORD followed by @<ISSUE NUMBER>__
KEYWORDS available are `close`, `closes`, `closed`, `fix`, `fixes`, `fixed`, `resolve`, `resolves`, `resolved`.
Please delete this text before creating your PR 

### Are there any further changes required?
No

### CODE REVIEW

File: examples/custom-gql-apis/src/hello_world.rs [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+use raphtory_core::core::utils::errors::GraphError;
+use raphtory_core::prelude::{CacheOps, GraphViewOps, ImportOps, NodeViewOps, PropertyAdditionOps};
+use raphtory_graphql::data::Data;
+use itertools::Itertools;
+use raphtory_graphql::model::plugins::query::Query;
+use raphtory_graphql::model::plugins::query_plugins::QueryPlugins;
+
+pub(crate) struct HelloWorld;
+
+impl<'a> Query<'a, QueryPlugins> for HelloWorld {
```
Comment (miratepuffin) [inline examples/custom-gql-apis/src/hello_world.rs:16]: Should QueryPlugins be singular i.e. QueryPlugin

File: examples/custom-gql-apis/src/python.rs [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+use raphtory_core::python::types::repr::{Repr, StructReprBuilder};
+use raphtory_graphql::python::server::server::PyGraphServer;
+use raphtory_graphql::python::server::take_server_ownership;
+use crate::hello_world::HelloWorld;
+
+#[pyfunction(name = "add_custom_gql_apis")]
+pub fn py_add_custom_gql_apis(
+    server: PyRefMut<PyGraphServer>,
+) -> Result<PyGraphServer, PyErr> {
+    let server = take_server_ownership(server)?;
```
Comment (miratepuffin) [inline examples/custom-gql-apis/src/python.rs:13]: Can this fail?

File: examples/custom-gql-apis/src/python.rs [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+    let server = server.register_query_plugins::<_, HelloWorld>("hello_world");
```
Comment (miratepuffin) [inline examples/custom-gql-apis/src/python.rs:14]: Can we add a mutation query as well as an example of how these are different.

File: raphtory-graphql/src/model/plugins/mutation_entry_point.rs [hunk 4]
```diff
@@ -0,0 +1,26 @@
+use crate::model::algorithms::RegisterFunction;
+use async_graphql::dynamic::Object;
+use dynamic_graphql::internal::{OutputTypeName, Register, Registry, ResolveOwned, TypeName};
+use itertools::Itertools;
+use std::{collections::HashMap, sync::MutexGuard};
+
+pub trait MutationEntryPoint<'a>:
+Register + TypeName + OutputTypeName + ResolveOwned<'a> + Sync
+{
+    fn lock_plugins() -> MutexGuard<'static, HashMap<String, RegisterFunction>>;
```
Comment (miratepuffin) [inline raphtory-graphql/src/model/plugins/mutation_entry_point.rs:10]: I assume this is called automatically and isn't up to the user to use?

File: examples/custom-gql-apis/src/python.rs (hunk 2 above)
Comment (ricopinazo) [reply examples/custom-gql-apis/src/python.rs:13]: yes if you do

```
server = ???
py_add_custom_gql_apis(server)
py_add_custom_gql_apis(server)
```
In the second call, the server would be already consumed. To be fair, we should change the server API to be mutable, as we did for the vector API

File: raphtory-graphql/src/model/plugins/mutation.rs [hunk 5]
```diff
@@ -0,0 +1,59 @@
+use crate::model::plugins::mutation_entry_point::MutationEntryPoint;
+use async_graphql::{
+    dynamic::{Field, FieldFuture, FieldValue, InputValue, Object, ResolverContext, TypeRef},
+    FieldResult,
+};
+use dynamic_graphql::internal::{Register, Registry};
+use futures_util::future::BoxFuture;
+use crate::model::plugins::mutation_plugins::MutationPlugins;
+
+pub trait Mutation<'a, A: MutationEntryPoint<'a> + 'static> {
```
Comment (ricopinazo) [inline raphtory-graphql/src/model/plugins/mutation.rs:10]: this trait looks the same as the Query trait. Can't we have a Function trait with an apply function that can be used for both mutations and queries?

File: raphtory-graphql/src/model/plugins/mutation_entry_point.rs [hunk 6]
```diff
@@ -0,0 +1,26 @@
+use crate::model::algorithms::RegisterFunction;
+use async_graphql::dynamic::Object;
+use dynamic_graphql::internal::{OutputTypeName, Register, Registry, ResolveOwned, TypeName};
+use itertools::Itertools;
+use std::{collections::HashMap, sync::MutexGuard};
+
+pub trait MutationEntryPoint<'a>:
```
Comment (ricopinazo) [inline raphtory-graphql/src/model/plugins/mutation_entry_point.rs:7]: This trait also seems to be exactly the same as the QueryEntryPoint. Maybe we can also have a PluginEntryPoint and implement it for both MutationPlugins and QueryPlugins

File: raphtory-graphql/src/model/plugins/mutation_entry_point.rs (hunk 4 above)
Comment (ricopinazo) [reply raphtory-graphql/src/model/plugins/mutation_entry_point.rs:10]: yeah it's only for internal use, maybe we should do this trait pub(crate), don't think anyone should be using it outside

**Review by miratepuffin — APPROVED** [review]

**Pull request merged**